routing/routing_tests/index_graph_test.cpp

```diff
@@ -0,0 +1,219 @@
+{
+  auto it = m_features.find(featureId);
+  if (it == m_features.end())
+  {
+    ASSERT(false, ("Can't find feature", featureId));
+    return 0.0;
+  }
+  vector<m2::PointD> const & points = it->second;
+
+  uint32_t const start = min(pointFrom, pointTo);
+  uint32_t const finish = max(pointFrom, pointTo);
+  ASSERT_LESS(finish, points.size(), ());
+
+  double result = 0.0;
+  for (uint32_t i = start; i < finish; ++i)
+    result += points[i].Length(points[i + 1]);
+
+  return result;
+}
+
+double TestGeometry::CalcHeuristic(FSegId from, FSegId to) const
+{
+  return GetPoint(from).Length(GetPoint(to));
+}
+
+void TestGeometry::AddRoad(uint32_t featureId, vector<m2::PointD> const & points)
+{
+  auto it = m_features.find(featureId);
+  if (it != m_features.end())
+  {
+    ASSERT(false, ("Already contains feature", featureId));
+    return;
+  }
+
+  m_features[featureId] = points;
+}
+
+Joint MakeJoint(vector<FSegId> const & points)
+{
+  Joint joint;
+  for (auto const & point : points)
+  {
+    joint.AddEntry(point);
+  }
+  return joint;
+}
+
+void CheckRoute(IndexGraph & graph, FSegId const & start, FSegId const & finish,
+                size_t expectedLength)
+{
+  LOG(LINFO, ("Check route", start.GetFeatureId(), ",", start.GetSegId(), "=>",
+              finish.GetFeatureId(), ",", finish.GetSegId()));
+
+  AStarAlgorithm<IndexGraph> algorithm;
+  RoutingResult<JointId> routingResult;
+
+  AStarAlgorithm<IndexGraph>::Result const resultCode = algorithm.FindPath(
+      graph, graph.InsertJoint(start), graph.InsertJoint(finish), routingResult, {}, {});
+  vector<FSegId> const & fsegs = graph.RedressRoute(routingResult.path);
+
+  TEST_EQUAL(resultCode, AStarAlgorithm<IndexGraph>::Result::OK, ());
+  TEST_EQUAL(fsegs.size(), expectedLength, ());
+}
+
+uint32_t AbsDelta(uint32_t v0, uint32_t v1) { return v0 > v1 ? v0 - v1 : v1 - v0; }
+}  // namespace
+
+namespace routing_test
+{
+//            R1:
+//
+//            -2
+//            -1
+//  R0: -2 -1  0  1  2
+//             1
+//             2
+//
+UNIT_TEST(FindPathCross)
+{
+  unique_ptr<TestGeometry> geometry = make_unique<TestGeometry>();
+  geometry->AddRoad(0, {{-2.0, 0.0}, {-1.0, 0.0}, {0.0, 0.0}, {1.0, 0.0}, {2.0, 0.0}});
+  geometry->AddRoad(1, {{0.0, -2.0}, {-1.0, 0.0}, {0.0, 0.0}, {0.0, 1.0}, {0.0, 2.0}});
+
+  IndexGraph graph(move(geometry));
+
+  graph.Export({MakeJoint({{0, 2}, {1, 2}})});
+
+  vector<FSegId> points;
+  for (uint32_t i = 0; i < 5; ++i)
+  {
+    points.emplace_back(0, i);
+    points.emplace_back(1, i);
+  }
+
+  for (auto const & start : points)
+  {
+    for (auto const & finish : points)
+    {
+      uint32_t expectedLength;
+      if (start.GetFeatureId() == finish.GetFeatureId())
+        expectedLength = AbsDelta(start.GetSegId(), finish.GetSegId()) + 1;
+      else
+        expectedLength = AbsDelta(start.GetSegId(), 2) + AbsDelta(finish.GetSegId(), 2) + 1;
```

Comment:
Why "+ 1". It looks like the length of the route should be the sum of lengths segments. Right?